plugin/storage/badger/spanstore/reader.go

```diff
@@ -0,0 +1,488 @@
+type TraceReader struct {
+	store *badger.DB
+	cache *CacheStore
+}
+
+// NewTraceReader returns a TraceReader with cache
+func NewTraceReader(db *badger.DB, c *CacheStore) *TraceReader {
+	return &TraceReader{
+		store: db,
+		cache: c,
+	}
+}
+
+// getTraces enriches TraceIDs to Traces
+func (r *TraceReader) getTraces(traceIDs []model.TraceID) ([]*model.Trace, error) {
+	// Get by PK
+	traces := make([]*model.Trace, 0, len(traceIDs))
+	prefixes := make([][]byte, 0, len(traceIDs))
+
+	for _, traceID := range traceIDs {
+		prefixes = append(prefixes, createPrimaryKeySeekPrefix(traceID))
+	}
+
+	err := r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchSize = 10 // TraceIDs are not sorted, pointless to prefetch large amount of values
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		val := []byte{}
+		for _, prefix := range prefixes {
+			spans := make([]*model.Span, 0, 4) // reduce reallocation requirements by defining some initial length
+
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				// Add value to the span store (decode from JSON / defined encoding first)
+				// These are in the correct order because of the sorted nature
+				item := it.Item()
+				val, err := item.ValueCopy(val)
+				if err != nil {
+					return err
+				}
+
+				sp := model.Span{}
+				switch item.UserMeta() & 0x0F {
+				case jsonEncoding:
+					if err := json.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				case protoEncoding:
+					if err := proto.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				default:
+					return fmt.Errorf("Unknown encoding type: %04b", item.UserMeta()&0x0F)
+				}
+				spans = append(spans, &sp)
+			}
+			trace := &model.Trace{
+				Spans: spans,
+			}
+			traces = append(traces, trace)
+		}
+		return nil
+	})
+
+	// TODO Do the Unmarshal here so we can release the transaction earlier (we would pay with extra allocations..)
+
+	return traces, err
+
+}
+
+// GetTrace takes a traceID and returns a Trace associated with that traceID
+func (r *TraceReader) GetTrace(traceID model.TraceID) (*model.Trace, error) {
+	traces, err := r.getTraces([]model.TraceID{traceID})
+	if err != nil {
+		return nil, err
+	} else if len(traces) == 1 {
+		return traces[0], nil
+	}
+
+	return nil, nil
+}
+
+func createPrimaryKeySeekPrefix(traceID model.TraceID) []byte {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(spanKeyPrefix)
+	binary.Write(buf, binary.BigEndian, traceID.High)
+	binary.Write(buf, binary.BigEndian, traceID.Low)
+	return buf.Bytes()
+}
+
+// GetServices fetches the sorted service list that have not expired
+func (r *TraceReader) GetServices() ([]string, error) {
+	return r.cache.GetServices()
+}
+
+// GetOperations fetches operations in the service and empty slice if service does not exists
+func (r *TraceReader) GetOperations(service string) ([]string, error) {
+	return r.cache.GetOperations(service)
```

Comment:
I noticed an issue with this implementation that in case of initializing badger factory by reloading from a pre-existing directory, the operations list does not get populated with services from pre-existing traces even though the traces are accessible by making a direct REST API call.
Probably the cause is that you are relying on the cache which is not pre-populated.

Reply:
NewCacheStore() calls prefillCaches()

Reply:
The unit test that does persisting (mentioned in other comment) also seems to indicate the caches are prefilled correctly.

I should add this type unit tests to the suite.

Reply:
I took a backup of Badger db and restored it to a location that was passed as key and value dir to the Jaeger standalone run. I noticed the loadServices() method does find services present in traces of the backup but finds its TTL as 0. This causes the services to be deleted during GetServices() as your logic finds it as expired. I did take a look at your test and somehow TTL in this case is 72 hours ahead of now time so the test passes.
I am not sure why services loaded from a pre-existing Badger db has 0 TTL. Do you know what might be causing this?

For now, commenting the delete line in GetServices() helps populate the services on UI- similarly for GetOperations(). Is there a danger in this approach? Why do you perform the check (v > t) in GetServices?

PS: My use case demands that backed up data should be accessible always.

Reply:
I can't replicate that behavior, in my testing I see:

```
Key: key="\x81service-p\x00\x05pSgE\x89\xbd\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01", version=1, meta=40, 1531137600, time now: 1530878401
Key: key="\x82service-poperation-p\x00\x05pSgE\x89\xbd\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01", version=1, meta=40, 1531137600, time now: 1530878401
```

That is, the services do expire later than the current time (as they should). The reason for that ``v > t`` is that we don't return ghost services or operations. That would be bad in terms of usability if one has a lot of services / operations that keep changing. 

If you need backups that do not expire ever, then I guess TTL is the wrong approach for you. What you'd want is something closer to a snapshot + purge of snapshotted data (with no TTL, such that you would manually take care of retention policies).

Certainly doable, but maybe it is a bit of feature creep for this PR?